tokio-executor/src/park.rs

```diff
@@ -237,37 +238,56 @@ impl Inner {
     fn park(&self, timeout: Option<Duration>) -> Result<(), ParkError> {
         // If currently notified, then we skip sleeping. This is checked outside
         // of the lock to avoid acquiring a mutex if not necessary.
-        match self.state.compare_and_swap(NOTIFY, IDLE, Ordering::SeqCst) {
-            NOTIFY => return Ok(()),
-            IDLE => {},
-            _ => unreachable!(),
+        if self.state.compare_exchange(NOTIFY, IDLE, SeqCst, SeqCst).is_ok() {
+            return Ok(());
+        }
+
+        // If the duration is zero, then there is no need to actually block
+        if let Some(ref dur) = timeout {
+            if *dur == Duration::from_millis(0) {
+                return Ok(());
+            }
         }
 
         // The state is currently idle, so obtain the lock and then try to
         // transition to a sleeping state.
         let mut m = self.mutex.lock().unwrap();
 
         // Transition to sleeping
-        match self.state.compare_and_swap(IDLE, SLEEP, Ordering::SeqCst) {
-            NOTIFY => {
+        match self.state.compare_exchange(IDLE, SLEEP, SeqCst, SeqCst) {
+            Ok(_) => {}
+            Err(NOTIFY) => {
                 // Notified before we could sleep, consume the notification and
                 // exit
-                self.state.store(IDLE, Ordering::SeqCst);
+                self.state.store(IDLE, SeqCst);
```

Comment:
This seems problematic for the same reason stated in the issue.

Given that this is intended to acquire all memory from `unpark` threads, it needs to establish a "happens-before" relationship with https://github.com/tokio-rs/tokio/blob/e433282cf15642b277f5fd6eabb94b1b68ce7f59/tokio-executor/src/park.rs#L303.

I do not think that this is done (though, again, MFENCE on x86 will work).